Automatically wrap autograd.grad with allow_in_graph

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #169415
* __->__ #169389

In this PR, we add support to wrap autograd.grad in allow_in_graph inside TorchDynamo. We can only do this when:
1) inputs to the graph region don't have external grad_fn or not a leaf node


I think we don't need to detect if the loss is detached in the output because aotautograd will error later with similar error message as eager mode. 

Fixes: https://github.com/pytorch/pytorch/issues/167729

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo

tugsbayasgalan committed:
Automatically wrap autograd.grad with allow_in_graph

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Automatically wrap autograd.grad with allow_in_graph"


In this PR, we add support to wrap autograd.grad in allow_in_graph inside TorchDynamo. We can only do this when:
1) inputs to the graph region don't have external grad_fn or not a leaf node
2) inputs to torch.autograd.grad region are not leaf node. 


Fixes: https://github.com/pytorch/pytorch/issues/167729

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Automatically wrap autograd.grad with allow_in_graph"


In this PR, we add support to wrap autograd.grad in allow_in_graph inside TorchDynamo. We can only do this when:
1) inputs to the graph region don't have external grad_fn or not a leaf node
2) inputs to torch.autograd.grad region are not leaf node. 


Fixes: https://github.com/pytorch/pytorch/issues/167729

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Automatically wrap autograd.grad with allow_in_graph"


In this PR, we add support to wrap autograd.grad in allow_in_graph inside TorchDynamo. We can only do this when:
1) inputs to the graph region don't have external grad_fn or not a leaf node
2) inputs to torch.autograd.grad region are not leaf node. 


Fixes: https://github.com/pytorch/pytorch/issues/167729

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Automatically wrap autograd.grad with allow_in_graph"


In this PR, we add support to wrap autograd.grad in allow_in_graph inside TorchDynamo. We can only do this when:
1) inputs to the graph region don't have external grad_fn or not a leaf node
2) inputs to torch.autograd.grad region are not leaf node. 


I think we don't need to detect if the loss is detached in the output because aotautograd will error later with similar error message as eager mode. 

Fixes: https://github.com/pytorch/pytorch/issues/167729

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

anijain2305 commented on torch/_dynamo/output_graph.py:
```diff
@@ -2070,6 +2070,49 @@ def run_compiler_collective(self) -> None:
+            ):
+                var = self.input_source_to_var[grapharg.source]
+                if isinstance(var, TensorVariable) and var.has_grad_fn:
+                    raise RuntimeError(
```
The error category is wrong. I think we should raise `unimplemented` with hints as `USER_ERROR`.

anijain2305 commented on test/dynamo/test_fwd_loss_bwd.py:
```diff
@@ -0,0 +1,253 @@
+            grads = torch.autograd.grad(loss, params)
+            return loss.detach(), grads[0].sum(), grads[1].sum()
+
+        cnt = CompileCounterWithBackend("aot_eager")
```
Use `AOTEagerAndRecordGraphs` backend and do an ExpectedInline on aot_eager as well. This will show that aot dispatcher is tracing through the autograd.grad.

anijain2305 reviewed (changes requested):
Have a few comments.

tugsbayasgalan committed:
Update on "Automatically wrap autograd.grad with allow_in_graph"


In this PR, we add support to wrap autograd.grad in allow_in_graph inside TorchDynamo. We can only do this when:
1) inputs to the graph region don't have external grad_fn or not a leaf node


I think we don't need to detect if the loss is detached in the output because aotautograd will error later with similar error message as eager mode. 

Fixes: https://github.com/pytorch/pytorch/issues/167729

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Automatically wrap autograd.grad with allow_in_graph"


In this PR, we add support to wrap autograd.grad in allow_in_graph inside TorchDynamo. We can only do this when:
1) inputs to the graph region don't have external grad_fn or not a leaf node


I think we don't need to detect if the loss is detached in the output because aotautograd will error later with similar error message as eager mode. 

Fixes: https://github.com/pytorch/pytorch/issues/167729

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Automatically wrap autograd.grad with allow_in_graph"


In this PR, we add support to wrap autograd.grad in allow_in_graph inside TorchDynamo. We can only do this when:
1) inputs to the graph region don't have external grad_fn or not a leaf node


I think we don't need to detect if the loss is detached in the output because aotautograd will error later with similar error message as eager mode. 

Fixes: https://github.com/pytorch/pytorch/issues/167729

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

ezyang commented on torch/_dynamo/output_graph.py:
```diff
@@ -2070,6 +2075,41 @@ def run_compiler_collective(self) -> None:
+                            "Compiled function receives an input with external grad_fn "
+                            "(created by operations outside the compiled region). "
+                            "When torch.autograd.grad is used, all inputs must be leaf tensors "
+                            "because the autograd graph cannot extend beyond the compiled region."
```
Because this is an overly conservative check, I think it's important for the explanation to call this out. And you may want to preemptively file an issue for a less conservative check. (In particular, it is safe to put the grad in if the non-leaf requires grad True input is NOT reachable from the autograd graph of the variables we differentiated.)

ezyang commented on torch/_dynamo/variables/torch.py:
```diff
@@ -1423,6 +1423,32 @@ def handle_check(self, tx: "InstructionTranslator", *args, **kwargs):
+        def handle_autograd_grad(self, tx: "InstructionTranslator", *args, **kwargs):
+            # Only allow autograd.grad in fullgraph=True mode.
+            # In non-fullgraph mode, graph breaks can cause tensors with grad_fn
+            # to become graph inputs, which breaks fake tensor tracing.
```
You probably want to flag this (the non graph break behavior) too.
